## Installation

1. If you haven't already installed Python3, get it from [Python.org](https://www.python.org/downloads/). **Tensorflow does not yet work with Python 3.7, so you _must_ get Python 3.6.** See https://github.com/tensorflow/tensorflow/issues/20517 for updates on 3.7 support.
1. If you haven't already installed Jupyter Notebook, run `python3 -m pip install jupyter`
1. In Terminal, cd to the folder in which you downloaded this file and run `jupyter notebook`. This should open up a page in your web browser that shows all of the files in the current directory, so that you can open this file. You will need to leave this Terminal window up and running and use a different one for the rest of the instructions.
1. Install the Gensim word2vec Python implementation: `pip3 install --upgrade gensim`
1. Get the trained model (1billion_word_vectors.zip) from me via airdrop or flashdrive and put it in the same folder as the ipynb file, the folder in which you are running the jupyter notebook command.
1. Unzip the trained model file. You should now have three files in the folder (if zip created a new folder, move these files out of that separate folder into the same folder as the ipynb file):
    * 1billion_word_vectors
    * 1billion_word_vectors.syn1neg.npy
    * 1billion_word_vectors.wv.syn0.npy
1. If you didn't install keras last time, install it now
    1. Install the tensorflow machine learning library by typing the following into Terminal:
    `pip3 install --upgrade tensorflow`
    1. Install the keras machine learning library by typing the following into Terminal:
    `pip3 install keras`


## Extra Details -- Do Not Do This
This took awhile, which is why I'm giving you the trained file rather than having you do this. But just in case you're curious, here is how to create the trained model file.
1. Download the corpus of sentences from [http://www.statmt.org/lm-benchmark/1-billion-word-language-modeling-benchmark-r13output.tar.gz](http://www.statmt.org/lm-benchmark/1-billion-word-language-modeling-benchmark-r13output.tar.gz)
1. Unzip and unarchive the file: `tar zxf 1-billion-word-language-modeling-benchmark-r13output.tar.gz` 
1. Run the following Python code:
    ```
    from gensim.models import word2vec
    import os

    corpus_dir = '1-billion-word-language-modeling-benchmark-r13output/training-monolingual.tokenized.shuffled'
    sentences = word2vec.PathLineSentences(corpus_dir)
    model = word2vec.Word2Vec(sentences) # just use all of the default settings for now
    model.save('1billion_word_vectors')
    ```

## Documentation/Sources
* [https://radimrehurek.com/gensim/models/word2vec.html](https://radimrehurek.com/gensim/models/word2vec.html) for more information about how to use gensim word2vec in general
* [https://codekansas.github.io/blog/2016/gensim.html](https://codekansas.github.io/blog/2016/gensim.html) for information about using it to create embedding layers for neural networks.
* [https://machinelearningmastery.com/sequence-classification-lstm-recurrent-neural-networks-python-keras/](https://machinelearningmastery.com/sequence-classification-lstm-recurrent-neural-networks-python-keras/) for information on sequence classification with keras
* [https://blog.keras.io/using-pre-trained-word-embeddings-in-a-keras-model.html](https://blog.keras.io/using-pre-trained-word-embeddings-in-a-keras-model.html) for using pre-trained embeddings with keras (though the syntax they use for the model layers is different than most other tutorials I've seen).
* [https://keras.io/](https://keras.io/) Keras API documentation

## Load the trained word vectors

In [1]:
from gensim.models import word2vec

Load the trained model file into memory

In [2]:
wv_model = word2vec.Word2Vec.load('../../../Downloads/1billion_word_vectors/1billion_word_vectors')

Since we do not need to continue training the model, we can save memory by keeping the parts we need (the word vectors themselves) and getting rid of the rest of the model.

In [3]:
wordvec = wv_model.wv
del wv_model

## Exploration of word vectors
Now we can look at some of the relationships between different words.

Like [the gensim documentation](https://radimrehurek.com/gensim/models/word2vec.html), let's start with a famous example: king + woman - man

In [4]:
wordvec.most_similar(positive=['king', 'woman'], negative=['man'])

[('queen', 0.8407387733459473),
 ('monarch', 0.7541723251342773),
 ('prince', 0.7350203394889832),
 ('princess', 0.696908175945282),
 ('empress', 0.677180290222168),
 ('sultan', 0.6649758815765381),
 ('Chakri', 0.6451102495193481),
 ('goddess', 0.6439394950866699),
 ('ruler', 0.6275453567504883),
 ('kings', 0.6273428201675415)]

This next one does not work as well as I'd hoped, but it gets close. Maybe you can find a better example.

In [5]:
wordvec.most_similar(positive=['panda', 'eucalyptus'], negative=['bamboo'])

[('okapi', 0.7140712738037109),
 ('gibbon', 0.7034620046615601),
 ('koala', 0.697202742099762),
 ('cub', 0.6907659769058228),
 ('tortoise', 0.6886162757873535),
 ('beetle', 0.6859476566314697),
 ('salamander', 0.6855185031890869),
 ('psyllid', 0.6837549209594727),
 ('lynx', 0.6802860498428345),
 ('carnivore', 0.6794542670249939)]

[('hippopotamus', 0.5878522396087646),
 ('coyote', 0.5614131093025208),
 ('bushmeat', 0.5552693605422974),
 ('dolphin', 0.5539993643760681),
 ('bobcat', 0.5522228479385376),
 ('cow', 0.5484085083007812),
 ('dingo', 0.5420719385147095),
 ('chimpanzee', 0.5386090278625488),
 ('warthog', 0.5322991609573364),
 ('whitetail', 0.5261341333389282)]

Which one of these is not like the others?

In [8]:
wordvec.doesnt_match(['red', 'purple', 'laptop', 'turquoise', 'ruby'])

'laptop'

How far apart are different words?

In [9]:
wordvec.distances('laptop', ['computer', 'phone', 'rabbit'])

array([0.205414  , 0.36557418, 0.6597437 ], dtype=float32)

Let's see what one of these vectors actually looks like.

In [10]:
wordvec['textbook']

array([ 0.50756323, -2.8890731 ,  0.9743826 , -0.60089743, -0.23762947,
       -2.324566  , -0.64634913, -0.66476715, -2.3432739 ,  1.4446437 ,
       -0.15542823,  1.8248576 ,  1.1309539 , -0.21071543, -0.82512087,
       -0.2773584 , -0.1973424 , -0.5337731 ,  2.1143918 ,  1.0673765 ,
       -0.2341243 ,  1.5292411 ,  0.66977274,  1.1214821 , -0.57710004,
       -0.02504024,  0.6074397 ,  0.19416903, -1.1265849 , -0.6618393 ,
        1.7525213 ,  1.6232891 , -0.3886833 , -1.1867149 ,  0.45511633,
        1.4240934 , -0.87929034, -1.8920534 ,  2.6986032 , -0.5277589 ,
        2.1202435 ,  0.62670445,  1.0352231 ,  1.4998924 ,  2.5809426 ,
        0.74698585, -0.07757699, -0.67074645,  1.6887746 , -0.22081567,
        1.2107906 ,  0.16741815,  3.3496742 ,  1.1832954 ,  0.4423463 ,
        0.04771314, -0.14557275, -1.3345221 ,  1.3236852 ,  2.0154989 ,
       -0.6510446 ,  0.21808812, -0.31578887, -1.822629  ,  0.8436349 ,
       -1.1500564 ,  1.24044   , -2.6430037 ,  1.0617311 ,  1.20

What other methods are available to us?

In [11]:
help(wordvec)

Help on Word2VecKeyedVectors in module gensim.models.keyedvectors object:

class Word2VecKeyedVectors(WordEmbeddingsKeyedVectors)
 |  Mapping between words and vectors for the :class:`~gensim.models.Word2Vec` model.
 |  Used to perform operations on the vectors such as vector lookup, distance, similarity etc.
 |  
 |  Method resolution order:
 |      Word2VecKeyedVectors
 |      WordEmbeddingsKeyedVectors
 |      BaseKeyedVectors
 |      gensim.utils.SaveLoad
 |      builtins.object
 |  
 |  Methods defined here:
 |  
 |  get_keras_embedding(self, train_embeddings=False)
 |      Get a Keras 'Embedding' layer with weights set as the Word2Vec model's learned word embeddings.
 |      
 |      Parameters
 |      ----------
 |      train_embeddings : bool
 |          If False, the weights are frozen and stopped from being updated.
 |          If True, the weights can/will be further trained/updated.
 |      
 |      Returns
 |      -------
 |      `keras.layers.Embedding`
 |          Embedd

# Optional Exercise: Explore Word Vectors
What other interesting relationship can you find, using the methods used in the examples above or anything you find in the help message?

In [12]:
wordvec.most_similar(positive=['giraffe'], negative=['tall']) # girrafe thats not tall is a hippo

[('hippopotamus', 0.5878522396087646),
 ('coyote', 0.5614131093025208),
 ('bushmeat', 0.5552693605422974),
 ('dolphin', 0.5539993643760681),
 ('bobcat', 0.5522228479385376),
 ('cow', 0.5484085083007812),
 ('dingo', 0.5420719385147095),
 ('chimpanzee', 0.5386090278625488),
 ('warthog', 0.5322991609573364),
 ('whitetail', 0.5261341333389282)]

In [13]:
wordvec.most_similar(positive=['giraffe', 'short']) # short giraffe is a zebra

[('zebra', 0.69512939453125),
 ('rabbit', 0.6831350326538086),
 ('hedgehog', 0.6788915395736694),
 ('caterpillar', 0.6777664422988892),
 ('pelican', 0.6744252443313599),
 ('hippo', 0.6698236465454102),
 ('frog', 0.6694817543029785),
 ('turtle', 0.6676884293556213),
 ('gecko', 0.6632249355316162),
 ('llama', 0.660350501537323)]

## Using the word vectors in an embedding layer of a Keras model

In [14]:
from keras.models import Sequential
import numpy

Using TensorFlow backend.


You may have noticed in the help text for wordvec that it has a built-in method for converting into a Keras embedding layer.

In [15]:
test_embedding_layer = wordvec.get_keras_embedding()
test_embedding_layer.input_length = 1

In [16]:
embedding_model = Sequential()
embedding_model.add(test_embedding_layer)

But how do we actually use this? If you look at the [Keras Embedding Layer documentation](https://keras.io/layers/embeddings/) you might notice that it takes numerical input, not strings. How do we know which number corresponds to a particular word? In addition to having a vector, each word has an index:

In [17]:
wordvec.vocab['python'].index

30438

Let's see if we get the same vector from the embedding layer as we get from our word vector object.

In [18]:
wordvec['python']

array([-1.1750487e+00,  2.3066440e-04, -6.0706180e-01, -1.1156354e+00,
       -1.0580894e+00, -2.7154784e+00, -3.6140988e+00, -1.0810910e+00,
        1.1234255e+00, -7.7326834e-01, -1.3322397e+00,  9.2905626e-02,
       -2.4488842e+00, -1.7817341e-01, -3.5459950e+00, -1.7320968e+00,
        1.9397168e+00, -6.3734710e-01,  2.3254216e+00, -1.3535864e+00,
       -1.4451812e-01, -2.4297442e+00,  1.5498929e+00,  8.1969726e-01,
        9.0982294e-01, -6.6116208e-01,  3.8905215e-01,  3.3855909e-01,
       -7.5454485e-01, -1.0352553e+00, -2.5936973e+00,  1.2103225e+00,
       -3.0236175e+00,  3.0580134e+00, -3.9140179e+00,  4.0223894e-01,
        1.7356061e+00,  9.0976155e-01,  2.0956397e-02,  2.0190549e+00,
        4.5332021e-01, -1.6634842e+00, -4.8180079e-01,  2.0414692e-01,
       -5.9267312e-01, -1.4182589e+00, -9.7301149e-01,  5.1611459e-01,
        2.0727324e+00,  2.0064230e+00, -7.5027935e-02, -1.1723986e+00,
       -8.6943096e-01,  1.7028141e+00,  2.2190344e+00,  9.3605727e-01,
      

In [19]:
embedding_model.predict(numpy.array([[30438]]))

array([[[-1.1750487e+00,  2.3066440e-04, -6.0706180e-01, -1.1156354e+00,
         -1.0580894e+00, -2.7154784e+00, -3.6140988e+00, -1.0810910e+00,
          1.1234255e+00, -7.7326834e-01, -1.3322397e+00,  9.2905626e-02,
         -2.4488842e+00, -1.7817341e-01, -3.5459950e+00, -1.7320968e+00,
          1.9397168e+00, -6.3734710e-01,  2.3254216e+00, -1.3535864e+00,
         -1.4451812e-01, -2.4297442e+00,  1.5498929e+00,  8.1969726e-01,
          9.0982294e-01, -6.6116208e-01,  3.8905215e-01,  3.3855909e-01,
         -7.5454485e-01, -1.0352553e+00, -2.5936973e+00,  1.2103225e+00,
         -3.0236175e+00,  3.0580134e+00, -3.9140179e+00,  4.0223894e-01,
          1.7356061e+00,  9.0976155e-01,  2.0956397e-02,  2.0190549e+00,
          4.5332021e-01, -1.6634842e+00, -4.8180079e-01,  2.0414692e-01,
         -5.9267312e-01, -1.4182589e+00, -9.7301149e-01,  5.1611459e-01,
          2.0727324e+00,  2.0064230e+00, -7.5027935e-02, -1.1723986e+00,
         -8.6943096e-01,  1.7028141e+00,  2.2190344

Looks good, right? But let's not waste our time when the computer could tell us definitively and quickly:

In [20]:
embedding_model.predict(numpy.array([[wordvec.vocab['python'].index]]))[0][0] == wordvec['python']

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True])

Now we have a way to turn words into word vectors with Keras layers. Yes! Time to get some data.

## The IMDB Dataset
The [IMDB dataset](https://keras.io/datasets/#imdb-movie-reviews-sentiment-classification) consists of movie reviews that have been marked as positive or negative. (There is also a built-in dataset of [Reuters newswires](https://keras.io/datasets/#reuters-newswire-topics-classification) that have been classified by topic.)

In [21]:
from keras.datasets import imdb
(x_train, y_train), (x_test, y_test) = imdb.load_data()

17465344/17464789 [==============================] - 12s 1us/step


It looks like our labels consists of 0 or 1, which makes sense for positive and negative.

In [22]:
print(y_train[0:9])
print(max(y_train))
print(min(y_train))

[1 0 0 1 0 0 1 0 1]
1
0


But x is a bit more trouble. The words have already been converted to numbers -- numbers that have nothing to do with the word embeddings we spent time learning!

In [23]:
x_train[0]

[1,
 14,
 22,
 16,
 43,
 530,
 973,
 1622,
 1385,
 65,
 458,
 4468,
 66,
 3941,
 4,
 173,
 36,
 256,
 5,
 25,
 100,
 43,
 838,
 112,
 50,
 670,
 22665,
 9,
 35,
 480,
 284,
 5,
 150,
 4,
 172,
 112,
 167,
 21631,
 336,
 385,
 39,
 4,
 172,
 4536,
 1111,
 17,
 546,
 38,
 13,
 447,
 4,
 192,
 50,
 16,
 6,
 147,
 2025,
 19,
 14,
 22,
 4,
 1920,
 4613,
 469,
 4,
 22,
 71,
 87,
 12,
 16,
 43,
 530,
 38,
 76,
 15,
 13,
 1247,
 4,
 22,
 17,
 515,
 17,
 12,
 16,
 626,
 18,
 19193,
 5,
 62,
 386,
 12,
 8,
 316,
 8,
 106,
 5,
 4,
 2223,
 5244,
 16,
 480,
 66,
 3785,
 33,
 4,
 130,
 12,
 16,
 38,
 619,
 5,
 25,
 124,
 51,
 36,
 135,
 48,
 25,
 1415,
 33,
 6,
 22,
 12,
 215,
 28,
 77,
 52,
 5,
 14,
 407,
 16,
 82,
 10311,
 8,
 4,
 107,
 117,
 5952,
 15,
 256,
 4,
 31050,
 7,
 3766,
 5,
 723,
 36,
 71,
 43,
 530,
 476,
 26,
 400,
 317,
 46,
 7,
 4,
 12118,
 1029,
 13,
 104,
 88,
 4,
 381,
 15,
 297,
 98,
 32,
 2071,
 56,
 26,
 141,
 6,
 194,
 7486,
 18,
 4,
 226,
 22,
 21,
 134,
 476,
 26,
 480,
 5

Looking at the help page for imdb, it appears there is a way to get the word back. Phew.

In [24]:
help(imdb)

Help on module keras.datasets.imdb in keras.datasets:

NAME
    keras.datasets.imdb - IMDB sentiment classification dataset.

FUNCTIONS
    get_word_index(path='imdb_word_index.json')
        Retrieves the dictionary mapping words to word indices.
        
        # Arguments
            path: where to cache the data (relative to `~/.keras/dataset`).
        
        # Returns
            The word index dictionary.
    
    load_data(path='imdb.npz', num_words=None, skip_top=0, maxlen=None, seed=113, start_char=1, oov_char=2, index_from=3, **kwargs)
        Loads the IMDB dataset.
        
        # Arguments
            path: where to cache the data (relative to `~/.keras/dataset`).
            num_words: max number of words to include. Words are ranked
                by how often they occur (in the training set) and only
                the most frequent words are kept
            skip_top: skip the top N most frequently occurring words
                (which may not be informative)

In [25]:
imdb_offset = 3
imdb_map = dict((index + imdb_offset, word) for (word, index) in imdb.get_word_index().items())
imdb_map[0] = 'PADDING'
imdb_map[1] = 'START'
imdb_map[2] = 'UNKNOWN'

1646592/1641221 [==============================] - 2s 1us/step


The knowledge about the initial indices came from [this stack overflow post](https://stackoverflow.com/questions/42821330/restore-original-text-from-keras-s-imdb-dataset) after I got gibberish when I tried to translate the first review, below. It looks coherent now!

In [38]:
' '.join([imdb_map[word_index] for word_index in x_train[0]])

{34704: 'fawn',
 52009: 'tsukino',
 52010: 'nunnery',
 16819: 'sonja',
 63954: 'vani',
 1411: 'woods',
 16118: 'spiders',
 2348: 'hanging',
 2292: 'woody',
 52011: 'trawling',
 52012: "hold's",
 11310: 'comically',
 40833: 'localized',
 30571: 'disobeying',
 52013: "'royale",
 40834: "harpo's",
 52014: 'canet',
 19316: 'aileen',
 52015: 'acurately',
 52016: "diplomat's",
 25245: 'rickman',
 6749: 'arranged',
 52017: 'rumbustious',
 52018: 'familiarness',
 52019: "spider'",
 68807: 'hahahah',
 52020: "wood'",
 40836: 'transvestism',
 34705: "hangin'",
 2341: 'bringing',
 40837: 'seamier',
 34706: 'wooded',
 52021: 'bravora',
 16820: 'grueling',
 1639: 'wooden',
 16821: 'wednesday',
 52022: "'prix",
 34707: 'altagracia',
 52023: 'circuitry',
 11588: 'crotch',
 57769: 'busybody',
 52024: "tart'n'tangy",
 14132: 'burgade',
 52026: 'thrace',
 11041: "tom's",
 52028: 'snuggles',
 29117: 'francesco',
 52030: 'complainers',
 52128: 'templarios',
 40838: '272',
 52031: '273',
 52133: 'zaniacs',

## Train our IMDB word vectors

In [162]:
train_sentences = [['PADDING'] + [imdb_map[word_index] for word_index in review] for review in x_train]
test_sentences = [['PADDING'] + [imdb_map[word_index] for word_index in review] for review in x_test]
# test_sentences

In [43]:
imdb_wv_model = word2vec.Word2Vec(train_sentences + test_sentences + ['UNKNOWN'], min_count=1)

In [53]:
imdb_wordvec = imdb_wv_model.wv
imdb_wv_model.wv.get_vector('snake')

imdb_wv_model.wv.vectors # list of word vectors

array([[ 1.6590145e+00,  1.4736452e+00, -6.8540913e-01, ...,
        -2.6535430e+00,  2.0356432e-01, -4.5510137e-04],
       [-4.1798127e-01,  1.0759944e-01, -1.3482510e+00, ...,
         1.0320209e-01,  3.1259465e-01, -7.6060146e-01],
       [ 2.2419794e+00,  7.4977160e-01, -1.7560135e+00, ...,
        -2.3932431e+00,  5.4664069e-01, -3.7890822e-01],
       ...,
       [-9.2387728e-02,  4.1493218e-02, -2.5499558e-02, ...,
         9.5577091e-03, -8.2462551e-03, -2.6214413e-02],
       [-7.5508989e-02,  3.7948288e-02, -1.8952144e-02, ...,
         2.5171066e-02,  5.5607390e-03,  2.3218333e-03],
       [-5.9921246e-02,  2.5725137e-02, -7.4301129e-03, ...,
         1.0143042e-02,  2.0036595e-02,  1.6822279e-03]], dtype=float32)

## Process the dataset
For this exercise, we're going to keep all inputs the same length (we'll see how to do variable-length later). This means we need to choose a maximum length for the review, cutting off longer ones and adding padding to shorter ones. What should we make the length? Let's understand our data.

In [30]:
lengths = [len(review) for review in x_train + x_test]
print('Longest review: {} Shortest review: {}'.format(max(lengths), min(lengths)))



Longest review: 2697 Shortest review: 70


2697 words! Wow. Well, let's see how many reviews would get cut off at a particular cutoff.

In [31]:
cutoff = 500
print('{} reviews out of {} are over {}.'.format(
    sum([1 for length in lengths if length > cutoff]), 
    len(lengths), 
    cutoff))

8485 reviews out of 25000 are over 500.


In [40]:
from keras.preprocessing import sequence
x_train_padded = sequence.pad_sequences(x_train, maxlen=cutoff)
x_test_padded = sequence.pad_sequences(x_test, maxlen=cutoff)


array([1, 0, 0, ..., 0, 1, 0])

## Classification without using the pre-trained word vectors

In [33]:
from keras.models import Sequential
from keras.layers import Embedding, Conv1D, Dense, Flatten
# x_train = list of words (review), y_train = positive or negative review

Model definition.

In [34]:
no_vector_model = Sequential()
no_vector_model.add(Embedding(input_dim=len(imdb_map), output_dim=100, input_length=cutoff)) # vectorizes the input
no_vector_model.add(Conv1D(filters=32, kernel_size=5, activation='relu'))
no_vector_model.add(Conv1D(filters=32, kernel_size=5, activation='relu'))
no_vector_model.add(Conv1D(filters=32, kernel_size=5, activation='relu'))
no_vector_model.add(Conv1D(filters=32, kernel_size=5, activation='relu'))
no_vector_model.add(Conv1D(filters=32, kernel_size=5, activation='relu'))
no_vector_model.add(Flatten())
no_vector_model.add(Dense(units=128, activation='relu'))
no_vector_model.add(Dense(units=1, activation='sigmoid')) # because at the end, we want one yes/no answer
no_vector_model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['binary_accuracy'])

Train the model. __This takes awhile. You might not want to re-run it.__

In [35]:
no_vector_model.fit(x_train_padded, y_train, epochs=1, batch_size=64)

Epoch 1/1
25000/25000 [==============================] - 137s 5ms/step - loss: 0.4220 - binary_accuracy: 0.7691


Assess the model. __This takes awhile. You might not want to re-run it.__

In [36]:
no_vector_scores = no_vector_model.evaluate(x_test_padded, y_test)
print('loss: {} accuracy: {}'.format(*no_vector_scores))

25000/25000 [==============================] - 37s 1ms/step
loss: 0.35417315168380736 accuracy: 0.85252


# Exercise: Use the word vectors in a full model
Using the knowledge about how the imdb dataset and the keras embedding layer represent words, as detailed above, define a model that uses the pre-trained word vectors from the imdb dataset rather than an embedding that keras learns as it goes along. You'll need to swap out the embedding layer and feed in different training data.

For any model that you try, take notes about the performance you see or anything you notice about the differences between each of them.

## Other optional exercises:
* Try using the 1billion vector word embeddings instead of the imdb vectors.
* Try changing different hyperparameters of the model.
* Try training with the reuters data.

In [98]:
custom_embedding_layer = imdb_wordvec.get_keras_embedding(train_embeddings=False)
print(custom_embedding_layer.input_dim)
print(len(imdb_map))
print(custom_embedding_layer.input_length)
custom_embedding_layer.input_length = cutoff
custom_embedding_layer

88591
88587
None


In [99]:
custom_embedding_model = Sequential()
custom_embedding_model.add(custom_embedding_layer)

In [105]:
imdb_wordvec.vocab['python'].index
imdb_wordvec.index2word[88590]

'W'

In [106]:
# making sure we get the same thing from wordvec as from layer:

custom_embedding_model.predict(numpy.array([[imdb_wordvec.vocab["W"].index]]))[0][0] == imdb_wordvec["W"]



array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True])

In [107]:
custom_embedding_model.add(Conv1D(filters=32, kernel_size=5, activation='relu'))
custom_embedding_model.add(Conv1D(filters=32, kernel_size=5, activation='relu'))
custom_embedding_model.add(Conv1D(filters=32, kernel_size=5, activation='relu'))
custom_embedding_model.add(Conv1D(filters=32, kernel_size=5, activation='relu'))
custom_embedding_model.add(Conv1D(filters=32, kernel_size=5, activation='relu'))
custom_embedding_model.add(Flatten())
custom_embedding_model.add(Dense(units=128, activation='relu'))
custom_embedding_model.add(Dense(units=1, activation='sigmoid'))
custom_embedding_model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['binary_accuracy'])


In [108]:
custom_embedding_scores = custom_embedding_model.evaluate(x_test_padded, y_test)
print('loss: {} accuracy: {}'.format(*custom_embedding_scores))

25000/25000 [==============================] - 34s 1ms/step
loss: 0.6960130726242065 accuracy: 0.49792


In [109]:
custom_embedding_model.fit(x_train_padded, y_train, epochs=1, batch_size=64)

Epoch 1/1
25000/25000 [==============================] - 90s 4ms/step - loss: 0.6976 - binary_accuracy: 0.5221


In [110]:
custom_embedding_scores = custom_embedding_model.evaluate(x_test_padded, y_test)
print('loss: {} accuracy: {}'.format(*custom_embedding_scores))

25000/25000 [==============================] - 38s 2ms/step
loss: 0.68263579246521 accuracy: 0.5348


In [158]:
from keras.models import load_model
custom_model = load_model('custom_embedding_model.h5') # trained for 20 epochs on bandersnatch... overfit very badly
# here is a line from training:
# fit line = custom_embedding_model.fit(x_train_padded, y_train, epochs=20, batch_size=64, validation_data=(x_test_padded, y_test))
# 25000/25000 [==============================] - 4s 163us/step - loss: 0.0205 - binary_accuracy: 0.9929 - val_loss: 1.2920 - val_binary_accuracy: 0.7893


In [160]:
custom_embedding_scores = custom_model.evaluate(x_test_padded, y_test)
print('loss: {} accuracy: {}'.format(*custom_embedding_scores))

25000/25000 [==============================] - 37s 1ms/step
loss: 1.29203198015213 accuracy: 0.78928


In [147]:
custom_model.predict(x_test_padded[0:1])

array([[0.0001494]], dtype=float32)

In [155]:
no_vector_model.predict(x_test_padded[0:1])

array([[0.337531]], dtype=float32)

In [150]:
y_test[0:1]

array([0])

In [152]:
custom_model.predict(x_test_padded[1:2])

array([[0.83751917]], dtype=float32)

In [156]:
no_vector_model.predict(x_test_padded[1:2])

array([[0.99991167]], dtype=float32)

In [154]:
y_test[1:2]

array([1])

overall seems to have learned (so it works)... although 78% from this model is not quite as good as the 85% from the no vector model (which only trained for one epoch...)